src/fabric/src/fabric_drop_seq.erl

```diff
@@ -0,0 +1,1020 @@
+                                    WorkersDict,
+                                    nil
+                                ),
+                                fabric_util:log_timeout(
+                                    DefunctWorkers,
+                                    "set_drop_seq"
+                                ),
+                                {error, timeout};
+                            {error, Reason} ->
+                                {error, Reason}
+                        end
+                    after
+                        rexi_monitor:stop(RexiMon)
+                    end
+            end
+    end.
+
+-spec calculate_drop_seqs([#shard{}], uuid_map(), peer_checkpoints(), shard_sync_history()) ->
+    peer_checkpoints().
+calculate_drop_seqs(Shards, UuidMap, PeerCheckpoints0, ShardSyncHistory) ->
+    PeerCheckpoints1 = substitute_splits(Shards, UuidMap, PeerCheckpoints0),
+    PeerCheckpoints2 = crossref(PeerCheckpoints1, ShardSyncHistory),
+    ShardSyncCheckpoints = latest_shard_sync_checkpoints(ShardSyncHistory),
+    maps:merge_with(fun merge_peers/3, PeerCheckpoints2, ShardSyncCheckpoints).
+
+handle_set_drop_seq_reply(ok, Worker, {Results0, Waiting}) ->
+    DropSeq = proplists:get_value(drop_seq, Worker#shard.opts),
+    [B, E] = Worker#shard.range,
+    BHex = couch_util:to_hex_bin(<<B:32/integer>>),
+    EHex = couch_util:to_hex_bin(<<E:32/integer>>),
+    Range = <<BHex/binary, "-", EHex/binary>>,
+    Results1 = maps:merge_with(
+        fun(_Key, Val1, Val2) ->
+            maps:merge(Val1, Val2)
+        end,
+        Results0,
+        #{Range => #{Worker#shard.node => DropSeq}}
+    ),
+    if
+        Waiting == 0 ->
+            {stop, Results1};
+        true ->
+            {ok, {Results1, Waiting - 1}}
+    end;
+handle_set_drop_seq_reply(Error, _, _Acc) ->
+    {error, Error}.
+
+crossref(PeerCheckpoints0, ShardSyncHistory) ->
+    PeerCheckpoints1 = maps:fold(
+        fun({Range, Node}, {Uuid, Seq}, Acc1) ->
+            Others = maps:filter(
+                fun({R, S, _T}, _History) -> R == Range andalso S == Node end, ShardSyncHistory
+            ),
+            if
+                Seq == 0 ->
+                    %% propogate any 0 checkpoint as they would not be
+                    %% matched in shard sync history.
+                    maps:fold(
+                        fun({R, _S, T}, _History, Acc2) ->
+                            maps:merge_with(fun merge_peers/3, #{{R, T} => {<<>>, 0}}, Acc2)
+                        end,
+                        Acc1,
+                        Others
+                    );
+                true ->
+                    maps:fold(
+                        fun({R, _S, T}, History, Acc2) ->
+                            case
+                                lists:search(
+                                    fun({SU, SS, _TU, _TS}) ->
+                                        uuids_match([Uuid, SU]) andalso SS =< Seq
```

Comment:
Minor optimization. It would be faster if you would swap the order of checks.

```erlang
SS =< Seq andalso uuids_match([Uuid, SU])
```

Reply:
that's a good idea. it's the `SS =< Seq` condition that matters ("find the most recent checkpoint that includes the seq I care about"), the uuid check is a safeguard.